File: Assets/Bundles/Shared/Resources/TS/VoxelWorld/WorldAPI.lua
-- Compiled with unity-ts v2.1.0-75
local World = require("Shared/TS/VoxelWorld/World").World
local WorldAPI
do
	WorldAPI = setmetatable({}, {
		__tostring = function()
			return "WorldAPI"
		end,
	})
	WorldAPI.__index = WorldAPI
	function WorldAPI.new(...)
		local self = setmetatable({}, WorldAPI)
		return self:constructor(...) or self
	end
	function WorldAPI:constructor()
	end
	function WorldAPI:GetMainWorld()
		if self.world then
			return self.world
		end
		local voxelWorld = GameObject:Find("VoxelWorld"):GetComponent("VoxelWorld")
		self.world = World.new(voxelWorld)
		return self.world
	end
	WorldAPI.DefaultVoxelHealth = 10
	WorldAPI.ChildVoxelId = 32
end
return {
	WorldAPI = WorldAPI,
}
-- ----------------------------------
-- ----------------------------------

File: Assets/Bundles/Imports/Core/Core~/src/Shared/VoxelWorld/WorldAPI.ts
import { MathUtil } from "Shared/Util/MathUtil";
import { BlockHitDamageFunc } from "./BlockHitDamageFunc";
import { World } from "./World";

export class WorldAPI {
	private static world: World | undefined;
	public static DefaultVoxelHealth = 10;
	public static ChildVoxelId = 32;

	public static GetMainWorld(): World {
		if (this.world) {
			return this.world;
		}

		const voxelWorld = GameObject.Find("VoxelWorld").GetComponent<VoxelWorld>();
		this.world = new World(voxelWorld);
		return this.world;
	}

	public static GetVoxelPosition(worldPosition: Vector3) {
		return MathUtil.FloorVec(worldPosition);
	}

	public static BlockHitDamageFunc: BlockHitDamageFunc = (player, block, blockPos, breakBlockMeta) => {
		// if (BedWars.IsMatchServer()) {
		// 	// BW: disable breaking map blocks
		// 	if (block.itemType !== ItemType.BED) {
		// 		const wasPlacedByUser = BlockDataAPI.GetBlockData<boolean>(blockPos, "placedByUser");
		// 		if (!wasPlacedByUser) {
		// 			return 0;
		// 		}
		// 	}

		// 	// BW: dont allow breaking your own team's bed
		// 	const teamBlockId = BlockDataAPI.GetBlockData<string>(blockPos, "teamId");
		// 	if (teamBlockId !== undefined && teamBlockId === player.GetTeam()?.id) {
		// 		return 0;
		// 	}

		// 	const blockMeta = WorldAPI.GetMainWorld().GetBlockAt(blockPos).itemMeta;
		// 	if (
		// 		breakBlockMeta.extraDamageBlockArchetype !== BlockArchetype.NONE &&
		// 		blockMeta?.block?.blockArchetype === breakBlockMeta.extraDamageBlockArchetype
		// 	) {
		// 		damage += breakBlockMeta.extraDamage;
		// 	}

		// 	const breakSpeedState = TeamUpgradeUtil.GetUpgradeStateForPlayer(TeamUpgradeType.BREAK_SPEED, player);
		// 	if (breakSpeedState && breakSpeedState.currentUpgradeTier > 0) {
		// 		const breakSpeedValue = TeamUpgradeUtil.GetUpgradeTierForType(
		// 			TeamUpgradeType.BREAK_SPEED,
		// 			breakSpeedState.currentUpgradeTier,
		// 		).value;
		// 		damage *= 1 + breakSpeedValue / 100;
		// 	}
		// }

		return breakBlockMeta.damage;
	};
}
